### DISCUSSION TITLE
WiFi connection issues without explicitly setting WiFi mode

### DISCUSSION THREAD

**Discussion by jpalazz2** [PlummersSoftwareLLC/NightDriverStrip#173] [Q&A]
# Board Will Not Connect Without Explicitly Setting WiFi Mode

## TL;DR
My controller refused to connect to wifi until I added a line to explicitly set the wifi mode to Station (`WiFi.mode(WIFI_STA)`) somewhere in `setup` in `main.cpp`. I want to know whether or not I've done something silly and if not if you'd be open to me putting up a pull request to add this one liner.

# Reproduction Steps
1. Clone a fresh copy of `PlummersSoftwareLLC/NightDriverStrip`
2. Duplicate `secrets.example.h` as `secrets.h`
3. Update SSID/password settings in `secrets.h`
4. In `globals.h` set `ENABLE_WIFI` to 1 inside the configuration block for the DEMO config
5. `pio run -e demo -t upload`
6. Open the serial monitor

## Expected Behavior
The board connects to wifi and reports it's local IP address over the serial monitor

## Actual Behavior
The board attempts to connect to wifi and fails endlessly

## My Board
[Adafruit HUZZAH32 ESP32 Feather](https://learn.adafruit.com/adafruit-huzzah32-esp32-feather)

# My Debug Process
* Verify that the network I'm trying to connect to and credentials are valid by connecting from my laptop. Check.
* Verify that my board is able to actually pick up the SSID being broadcast using [this nifty little example sketch from arduino](https://create.arduino.cc/example/library/wifi_1_2_7/wifi_1_2_7%5Cexamples%5CScanNetworks/ScanNetworks/preview). The SSID is printed in the Serial monitor so my board can see the network.
* Write the [simplest program I can produce](#my-basic-network-code) that attempts to connect to the network, this succeeds and reports my local IP address.

At this point I started picking through the code and I observed that the only difference between my basic networking script and NightDriver is that the WiFi mode was not being set explicitly and relying on the default value. I added a quick print statement into the network loop that attempts connections to verify that the default mode was set to Station (and it was). Just to be thorough I decided to add a single line to `setup` which set the WiFi mode

<img width="288" alt="Screenshot 2022-12-08 at 12 45 46 AM" src="https://user-images.githubusercontent.com/6569083/206388914-5591c634-376d-4082-82b8-f22d984642d4.png">

and Voila!

<img width="374" alt="Screenshot 2022-12-08 at 12 09 49 AM" src="https://user-images.githubusercontent.com/6569083/206388986-64bab772-aa59-431d-9175-136dc805d51b.png">

So my question is: am I doing something silly that I'm just not noticing? If not how do you feel about me putting up a pull request that will explicitly set the WiFi mode when `ENABLE_WIFI` is set.

## My Basic Network Code
```
#include "Arduino.h"
#include "WiFi.h"

void setup()
{
  Serial.begin(115200);

  // Set WiFi to station mode and disconnect from an AP if it was previously connected
  WiFi.mode(WIFI_STA);
  WiFi.disconnect();
  delay(100);

  Serial.println("Setup done");
}

void loop()
{
  WiFi.begin("iot", "myssidpassword");
  Serial.println("Connecting to wifi...");
  while (WiFi.status() != WL_CONNECTED)
  {
    Serial.print('.');
    delay(1000);
  }

  Serial.println(WiFi.localIP());
}
```

**Comment by MatthewvanBuuren** [comment]
Its not set in any other place and AP is probably the most appropriate mode, I favour on being explicit so inside the 'ConnectToWiFi' function is probably the appropriate place.

**Comment by bungard** [comment]
This might be a larger issue, but i noticed that when I enabled "DEMO=1" in the build args..my esp32 won't connect to wifi (just stuck in a connect loop). 

The following config will connect to wifi, and report the IP. Adding `-DDEMO=1` to the build args..and it won't connect. 
```
[env:esp32doit-devkit-v1]
platform = espressif32
board = esp32doit-devkit-v1
monitor_speed = 115200
upload_port   = COM3
upload_speed  = 1500000
build_flags   = -DLEDSTRIP=1
                -std=gnu++17
                -Dregister=                         ; Sinister:  redefine 'register' so FastLED can use that keyword under C++17
                -Ofast 
```

For good measure, I added this to network.cpp..just to verify it wasn't bad credentials
```
else if (WiFi.status() == WL_CONNECT_FAILED)
            {
                Serial.println("Failed to connect to WIFI. Please verify credentials: ");
                Serial.println();
                Serial.print("SSID: ");
                Serial.println(cszSSID);
                Serial.print("Password: ");
                Serial.println(cszPassword);
                Serial.println();
            } 
```

**Reply by rbergen** [reply]
That's because WiFi is disabled in the DEMO setup. Specifically, `ENABLE_WIFI` is set to 0 in [this line in include/globals.h](https://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/cb4a3d093577c121c0e02855001d6c46b464a25e/include/globals.h#L241).

**Reply by bungard** [reply]
in my local checkout, i had enabled WIFI options in the `#if DEMO`  block
![image](https://user-images.githubusercontent.com/1907931/207625781-86038d7b-d51b-40ea-86e9-4be62ce9a213.png)

I'll jump back into it later, and try to dump out some other output for perspective.

**Reply by rbergen** [reply]
That obviously changes things. :) In that case it would indeed be interesting to understand better what's going on, here.

**Reply by bungard** [reply]
poking at some of the other projects, I saw that some of the WIFI ones defined `#define RESERVE_MEMORY  120000` (Or some number thereabout). 

Adding that to the DEMO block, seems to have fixed my issue. 
Looks like main.cpp is allocating all the memory it can for the LEDs.. (even if it doesn't need to?)


Here is some output when WIFI was working (w/ RESERVE_MEMORY defined)
``` 
(I) (PrintOutputHeader)(C1) Version 34: Wifi SSID: andthere2 - ESP32 Free Memory: 258084, PSRAM:0, PSRAM Free: 0
(I) (PrintOutputHeader)(C1) ESP32 Clock Freq : 240 MHz
...
(I) (setup)(C1) Starting SPIFFS...
(I) (setup)(C1) SPIFFS OK!
(W) (setup)(C1) Reserving 188 LED buffers for a total of 87232 bytes...
....
(C1) Setup complete - ESP32 Free Memory: 128532
...
Pass 0 of 10: Connecting to Wifi SSID: andthere2 - ESP32 Free Memory: 127832, PSRAM:0, PSRAM Free: 0
Connected to AP with BSSID: ....
```

Here is some output w/o RESERVE_MEMORY:
```
(I) (PrintOutputHeader)(C1) Version 34: Wifi SSID: andthere2 - ESP32 Free Memory: 288648, PSRAM:0, PSRAM Free: 0
(I) (PrintOutputHeader)(C1) ESP32 Clock Freq : 240 MHz
(I) (setup)(C1) Startup!
(I) (setup)(C1) Starting DebugLoopTaskEntry
(I) (DebugLoopTaskEntry)(C0) >> DebugLoopTaskEntry
(I)
(W) (setup)(C1) Reserving 44 LED buffers for a total of 153472 bytes...
(I) (setup)(C1) Adding LEDs to FastLED...
(I) (setup)(C1) Adding 1152 LEDs to FastLED.
...
(C1) Setup complete - ESP32 Free Memory: 99236
```

**Comment by davepl** [comment]
If you don’t need a bunch of buffers for LED WiFi, you can set MAX_BUFFERS in your config.  Set it to 1 or 2, not sure what is the lowest you can go.  
If you do need buffers but need more memory left available afterward, you can increase RESERVE_MEMORY instead, and that’s how much it will leave behind when it allocated the buffers.

Thanks!
Dave

**Reply by bungard** [reply]
Thanks @davepl - I hadn't caught the use of MAX_BUFFERS before..I'll give that a look
